File: examples/mini_lang/parser.py
"""Recursive-descent parser: tokens -> list of statements.

Grammar (precedence climbing by hand):

    program := (statement (SEP)*)*            SEP = ';' or newline
    statement := 'let' IDENT '=' expr | expr
    expr   := term (('+' | '-') term)*
    term   := factor (('*' | '/') factor)*
    factor := '-' factor | primary
    primary:= NUMBER | IDENT | '(' expr ')'
"""

from __future__ import annotations

from typing import List

from ast_nodes import BinOp, ExprStmt, LetStmt, Number, Stmt, UnaryOp, Var
from errors import ParseError
from tokens import Token, TokenKind


class Parser:
    def __init__(self, tokens: List[Token]) -> None:
        self.tokens = tokens
        self.i = 0

    def _peek(self) -> Token:
        return self.tokens[self.i]

    def _advance(self) -> Token:
        tok = self.tokens[self.i]
        self.i += 1
        return tok

    def _expect(self, kind: TokenKind) -> Token:
        tok = self._peek()
        if tok.kind is not kind:
            raise ParseError(f"expected {kind.value}, got {tok.kind.value} {tok.text!r}")
        return self._advance()

    def parse_program(self) -> List[Stmt]:
        stmts: List[Stmt] = []
        while self._peek().kind is not TokenKind.EOF:
            if self._peek().kind is TokenKind.SEMICOLON:
                self._advance()
                continue
            stmts.append(self._statement())
            nxt = self._peek().kind
            if nxt not in (TokenKind.SEMICOLON, TokenKind.EOF):
                tok = self._peek()
                raise ParseError(f"unexpected token {tok.text!r} after statement")
        return stmts

    def _statement(self) -> Stmt:
        if self._peek().kind is TokenKind.LET:
            self._advance()
            name = self._expect(TokenKind.IDENT).text
            self._expect(TokenKind.EQUALS)
            return LetStmt(name, self._expr())
        return ExprStmt(self._expr())

    def _expr(self):
        node = self._term()
        while self._peek().kind in (TokenKind.PLUS, TokenKind.MINUS):
            op = self._advance().text
            node = BinOp(op, node, self._term())
        return node

    def _term(self):
        node = self._factor()
        while self._peek().kind in (TokenKind.STAR, TokenKind.SLASH):
            op = self._advance().text
            node = BinOp(op, node, self._factor())
        return node

    def _factor(self):
        if self._peek().kind is TokenKind.MINUS:
            self._advance()
            return UnaryOp("-", self._factor())
        return self._primary()

    def _primary(self):
        tok = self._peek()
        if tok.kind is TokenKind.NUMBER:
            self._advance()
            return Number(float(tok.text))
        if tok.kind is TokenKind.IDENT:
            self._advance()
            return Var(tok.text)
        if tok.kind is TokenKind.LPAREN:
            self._advance()
            node = self._expr()
            self._expect(TokenKind.RPAREN)
            return node
        raise ParseError(f"unexpected token {tok.text!r}" if tok.text else "unexpected end of input")


def parse(tokens: List[Token]) -> List[Stmt]:
    return Parser(tokens).parse_program()

File: examples/mini_lang_rust/src/parser.rs
//! Recursive-descent parser: tokens -> statements. Mirrors examples/mini_lang/parser.py.

use crate::ast_nodes::{Expr, Stmt};
use crate::errors::{pyrepr, MiniLangError};
use crate::tokens::{Token, TokenKind};

pub struct Parser {
    tokens: Vec<Token>,
    i: usize,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Parser { tokens, i: 0 }
    }

    fn peek(&self) -> &Token {
        &self.tokens[self.i]
    }

    fn advance(&mut self) -> Token {
        let tok = self.tokens[self.i].clone();
        self.i += 1;
        tok
    }

    fn expect(&mut self, kind: TokenKind) -> Result<Token, MiniLangError> {
        let tok = self.peek();
        if tok.kind != kind {
            return Err(MiniLangError::Parse(format!(
                "expected {}, got {} {}",
                kind.name(),
                tok.kind.name(),
                pyrepr(&tok.text),
            )));
        }
        Ok(self.advance())
    }

    pub fn parse_program(&mut self) -> Result<Vec<Stmt>, MiniLangError> {
        let mut stmts = Vec::new();
        while self.peek().kind != TokenKind::Eof {
            if self.peek().kind == TokenKind::Semicolon {
                self.advance();
                continue;
            }
            stmts.push(self.statement()?);
            let nxt = self.peek().kind;
            if nxt != TokenKind::Semicolon && nxt != TokenKind::Eof {
                let tok = self.peek();
                return Err(MiniLangError::Parse(format!(
                    "unexpected token {} after statement",
                    pyrepr(&tok.text)
                )));
            }
        }
        Ok(stmts)
    }

    fn statement(&mut self) -> Result<Stmt, MiniLangError> {
        if self.peek().kind == TokenKind::Let {
            self.advance();
            let name = self.expect(TokenKind::Ident)?.text;
            self.expect(TokenKind::Equals)?;
            return Ok(Stmt::Let(name, self.expr()?));
        }
        Ok(Stmt::Expr(self.expr()?))
    }

    fn expr(&mut self) -> Result<Expr, MiniLangError> {
        let mut node = self.term()?;
        while matches!(self.peek().kind, TokenKind::Plus | TokenKind::Minus) {
            let op = self.advance().text.chars().next().unwrap();
            node = Expr::Binary(op, Box::new(node), Box::new(self.term()?));
        }
        Ok(node)
    }

    fn term(&mut self) -> Result<Expr, MiniLangError> {
        let mut node = self.factor()?;
        while matches!(self.peek().kind, TokenKind::Star | TokenKind::Slash) {
            let op = self.advance().text.chars().next().unwrap();
            node = Expr::Binary(op, Box::new(node), Box::new(self.factor()?));
        }
        Ok(node)
    }

    fn factor(&mut self) -> Result<Expr, MiniLangError> {
        if self.peek().kind == TokenKind::Minus {
            self.advance();
            return Ok(Expr::Unary(Box::new(self.factor()?)));
        }
        self.primary()
    }

    fn primary(&mut self) -> Result<Expr, MiniLangError> {
        let tok = self.peek().clone();
        match tok.kind {
            TokenKind::Number => {
                self.advance();
                Ok(Expr::Number(tok.text.parse::<f64>().unwrap()))
            }
            TokenKind::Ident => {
                self.advance();
                Ok(Expr::Var(tok.text))
            }
            TokenKind::LParen => {
                self.advance();
                let node = self.expr()?;
                self.expect(TokenKind::RParen)?;
                Ok(node)
            }
            _ => {
                if tok.text.is_empty() {
                    Err(MiniLangError::Parse("unexpected end of input".to_string()))
                } else {
                    Err(MiniLangError::Parse(format!(
                        "unexpected token {}",
                        pyrepr(&tok.text)
                    )))
                }
            }
        }
    }
}

pub fn parse(tokens: Vec<Token>) -> Result<Vec<Stmt>, MiniLangError> {
    Parser::new(tokens).parse_program()
}
